src/breakpoints.jl

```diff
@@ -1,26 +1,103 @@
-const _breakpoints = BreakpointRef[]
+const _breakpoints = AbstractBreakpoint[]
 breakpoints() = copy(_breakpoints)
 
-Base.getindex(bp::BreakpointRef) = bp.framecode.breakpoints[bp.stmtidx]
-function Base.setindex!(bp::BreakpointRef, isactive::Bool)
-    bp.framecode.breakpoints[bp.stmtidx] = BreakpointState(isactive, bp[].condition)
+function add_to_existing_framecodes(bp::AbstractBreakpoint)
+    for framecode in values(framedict)
+        add_breakpoint_if_match!(framecode, bp)
+    end
 end
-function toggle!(bp::BreakpointRef)
-    state = bp[]
-    bp.framecode.breakpoints[bp.stmtidx] = BreakpointState(!state.isactive, state.condition)
+
+function add_breakpoint_if_match!(framecode::FrameCode, bp::AbstractBreakpoint)
```

Comment:
Should we do this only for `BreakpointSignature`s? My biggest concern regards my last comment in https://github.com/JuliaDebug/JuliaInterpreter.jl/pull/256#issuecomment-478649540: what to do about cases where users specified a breakpoint by file/line and then revised the file, thus changing the line numbers. Perhaps a more robust approach would be to convert BreakpointFileLocations to BreakpointSignatures as soon as they resolve?

Reply:
Hm, perhaps a `BreakpointFileLocation` should only be able to bind once to a method and if the method is reevaluated you need to re-add the breakpoint.  It would be similar to the current situation except that you can now do "late binding", i.e. add a breakpoint to a file:line before it is included.

Reply:
But the same thing sort of applies to a breakpoint to a method or signature with a line number. Should those also only trigger once?

Reply:
Yes, I'm fine with triggering only once. In Juno one might imagine seeing the breakpoint vanish when you save the file, so it should be very easy for users to compensate.

Reply:
Hm, how would Juno know the breakpoint is invalid? The breakpoint instance will still be there in `_breakpoints`, it is just that when we interpret the method, another `FrameCode` will be executed.

FWIW, I think Juno should implement location based breakpoints on a level higher than here. In IDEs the breakpoint typically to follow along with edits, so if you insert a line above a breakpoint, the breakpoint will move down a line (together with the rest of the text). Clearly, we cannot `breakpoint(file, line)` and `remove` every time a user enter a line, so we need to "apply" the current set of line breakpoints when the debugger is started, at the locations they currently are.

cc @pfitzseb

Reply:
Revise could fire a callback notifying Juno of deleted methods. Though I guess `revise()` will run when the user tries the next command, not when the user saves the file. So the appropriate spot may no longer be visible...

Reply:
It's also not only Revise that can redefine methods. Ctrl + enter in Juno or just re-including a file for example. Losing breakpoints because you reinclude a file would perhaps be annoying.